src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NativeInteger.cs

```diff
@@ -0,0 +1,946 @@
+                    AttributeTargets.Parameter |
+                    AttributeTargets.Property |
+                    AttributeTargets.ReturnValue;
+                Assert.Equal(expectedTargets, attributeUsage.ValidTargets);
+            });
+        }
+
+        [Fact]
+        public void AttributeFieldExists()
+        {
+            var source =
+@"public class Program
+{
+    public nint F;
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            CompileAndVerify(comp, symbolValidator: module =>
+            {
+                var type = module.ContainingAssembly.GetTypeByMetadataName("Program");
+                var member = type.GetMembers("F").Single();
+                var attributes = member.GetAttributes();
+                AssertNativeIntegerAttribute(attributes);
+                var attribute = GetNativeIntegerAttribute(attributes);
+                var field = attribute.AttributeClass.GetField("TransformFlags");
+                Assert.Equal("System.Boolean[]", field.TypeWithAnnotations.ToTestDisplayString());
+            });
+        }
+
+        private static TypeDefinition GetTypeDefinitionByName(MetadataReader reader, string name)
+        {
+            return reader.GetTypeDefinition(reader.TypeDefinitions.Single(h => reader.StringComparer.Equals(reader.GetTypeDefinition(h).Name, name)));
+        }
+
+        private static string GetAttributeConstructorName(MetadataReader reader, CustomAttributeHandle handle)
+        {
+            return reader.Dump(reader.GetCustomAttribute(handle).Constructor);
+        }
+
+        private static CustomAttribute GetAttributeByConstructorName(MetadataReader reader, CustomAttributeHandleCollection handles, string name)
+        {
+            return reader.GetCustomAttribute(handles.FirstOrDefault(h => GetAttributeConstructorName(reader, h) == name));
+        }
+
+        private static void AssertAttributes(MetadataReader reader, CustomAttributeHandleCollection handles, params string[] expectedNames)
+        {
+            var actualNames = handles.Select(h => GetAttributeConstructorName(reader, h)).ToArray();
+            AssertEx.SetEqual(actualNames, expectedNames);
+        }
+
+        private static void AssertNoNativeIntegerAttribute(ImmutableArray<CSharpAttributeData> attributes)
+        {
+            AssertAttributes(attributes);
+        }
+
+        private static void AssertNativeIntegerAttribute(ImmutableArray<CSharpAttributeData> attributes)
+        {
+            AssertAttributes(attributes, "System.Runtime.CompilerServices.NativeIntegerAttribute");
+        }
+
+        private static void AssertAttributes(ImmutableArray<CSharpAttributeData> attributes, params string[] expectedNames)
+        {
+            var actualNames = attributes.Select(a => a.AttributeClass.ToTestDisplayString()).ToArray();
+            AssertEx.SetEqual(actualNames, expectedNames);
```

Comment:
I'm surprised that some of these methods are necessary over just `Assert.Equal(IEnumerable<T>, IEnumerable<T>)` now? SetEqual makes it sound like we just want the same elements in any order, but determinism requires the attributes are always in the same order, right?

Reply:
You're right, the elements should be in a fixed order. I'll change to `AssertEx.Equal()` in a subsequent PR.

---
In reply to: [373706572](https://github.com/dotnet/roslyn/pull/41160#discussion_r373706572) [](ancestors = 373706572)